src/runtime/HalideBuffer.h

```diff
@@ -2048,33 +2059,29 @@ class Buffer {
             }
         }
 
-        bool innermost_strides_are_one = true;
+        *innermost_strides_are_one = true;
         if (dimensions > 0) {
             for (int i = 0; i < N; i++) {
-                innermost_strides_are_one &= (t[0].stride[i] == 1);
+                *innermost_strides_are_one &= (t[0].stride[i] == 1);
             }
         }
-
-        return innermost_strides_are_one;
     }
 
     template<typename Fn, typename... Args, int N = sizeof...(Args) + 1>
     void for_each_value_impl(Fn &&f, Args &&... other_buffers) const {
-        bool all_buffers_defined = data() != nullptr;
-        bool do_null_checks[] = {(all_buffers_defined &= (other_buffers.data() != nullptr))...};
-        (void)do_null_checks;
-        assert(all_buffers_defined && "Called for_each_value on an undefined Buffer");
         Buffer<>::for_each_value_task_dim<N> *t =
             (Buffer<>::for_each_value_task_dim<N> *)HALIDE_ALLOCA((dimensions() + 1) * sizeof(for_each_value_task_dim<N>));
         // Move the preparatory code into a non-templated helper to
         // save code size.
         const halide_buffer_t *buffers[] = {&buf, (&other_buffers.buf)...};
-        bool innermost_strides_are_one = Buffer<>::for_each_value_prep(t, buffers);
-
-        Buffer<>::for_each_value_helper(f, dimensions() - 1,
-                                        innermost_strides_are_one,
-                                        t,
-                                        data(), (other_buffers.data())...);
+        bool innermost_strides_are_one, buffer_is_zero_size;
+        Buffer<>::for_each_value_prep(t, buffers, &innermost_strides_are_one, &buffer_is_zero_size);
+        if (!buffer_is_zero_size) {
```

Comment:
This implementation seems OK, but I was thinking of a much simpler change when I suggested it:

    assert(data() != nullptr || number_of_elements() == 0);

(But with some way of checking other_buffers.data()... as well, which your previous commit did.)

Reply:
Actually reviewed it now. I think you could keep this approach, but not actually return buffer_is_zero_size, and just use it to skip the assert. It's safe to just let it run for_each_value_helper with `t = {0...}`. That would make this a smaller change.